libs/dyn/pattern_trie.go

```diff
@@ -0,0 +1,200 @@
+package dyn
+
+import "fmt"
+
+// PatternTrie is a trie data structure for storing and querying patterns.
+// It supports both exact matches and wildcard matches. You can insert [Pattern]s
+// into the trie and then query it to see if a given [Path] matches any of the
+// patterns.
+type PatternTrie struct {
+	root *trieNode
+}
+
+// trieNode represents a node in the pattern trie.
+// Each node in the array represents one or more of:
+// 1. An [AnyKey] component. This is the "*" wildcard which matches any map key.
+// 2. An [AnyIndex] component. This is the "[*]" wildcard which matches any array index.
+// 3. Multiple [Key] components. These are multiple static path keys for this this node would match.
+// 4. Multiple [Index] components. These are multiple static path indices for this this node would match.
+//
+// Note: It's valid for both anyKey and pathKey to be set at the same time.
+// For example, adding both "foo.*.bar" and "foo.bar" to a trie is valid.
+// Similarly, it's valid for both anyIndex and pathIndex to be set at the same time.
+// For example, adding both "foo[*].bar" and "foo[0]" to a trie is valid.
+//
+// Note: Setting both key (one of pathKey or anyKey) and index (one of pathIndex or anyIndex)
+// is not supported by the [PatternTrie.SearchPath] method. We don't perform validation for this
+// case because it's not expected to arise in practice where a field is either a map or an array,
+// but not both.
+type trieNode struct {
+	// If set this indicates the trie node is an anyKey node.
+	// Maps to the [AnyKey] component.
+	anyKey *trieNode
+
+	// Indicates the trie node is an anyIndex node.
+	// Maps to the [AnyIndex] component.
+	anyIndex *trieNode
+
+	// Set of strings which this trie node matches.
+	// Maps to the [Key] component.
+	pathKey map[string]*trieNode
+
+	// Set of indices which this trie node matches.
+	// Maps to the [Index] component.
+	pathIndex map[int]*trieNode
```

Comment:
I cannot think of any case where we would need to match exact index array. Can be dropped to simplify?

Reply:
Removed support for specific indices.